htlcswitch/resolution_store.go

```diff
@@ -0,0 +1,168 @@
+package htlcswitch
+
+import (
+	"bytes"
+	"io"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/contractcourt"
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// resBucketKey is used for the root level bucket that stores the
+	// CircuitKey -> ResolutionMsg mapping.
+	resBucketKey = []byte("resolution-store-bucket-key")
+)
+
+// resolutionStore contains ResolutionMsgs received from the contractcourt. The
+// Switch deletes these from the store when the underlying circuit has been
+// removed via DeleteCircuits. If the circuit hasn't been deleted, the Switch
+// will dispatch the ResolutionMsg to a link if this was a multi-hop HTLC or to
+// itself if the Switch initiated the payment.
+type resolutionStore struct {
+	backend kvdb.Backend
+}
+
+func newResolutionStore(db kvdb.Backend) *resolutionStore {
+	return &resolutionStore{
+		backend: db,
+	}
+}
+
+// addResolutionMsg persists a ResolutionMsg to the resolutionStore.
+func (r *resolutionStore) addResolutionMsg(
+	resMsg *contractcourt.ResolutionMsg) error {
+
+	// The outKey will be the database key.
+	outKey := &CircuitKey{
+		ChanID: resMsg.SourceChan,
+		HtlcID: resMsg.HtlcIndex,
+	}
+
+	var resBuf bytes.Buffer
+	if err := serializeResolutionMsg(&resBuf, resMsg); err != nil {
+		return err
+	}
+
+	err := kvdb.Update(r.backend, func(tx kvdb.RwTx) error {
+		resBucket, err := tx.CreateTopLevelBucket(resBucketKey)
+		if err != nil {
+			return err
+		}
+
+		return resBucket.Put(outKey.Bytes(), resBuf.Bytes())
+	}, func() {})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fetchAllResolutionMsg returns a slice of all stored ResolutionMsgs. This is
+// used by the Switch on start-up.
+func (r *resolutionStore) fetchAllResolutionMsg() (
+	[]*contractcourt.ResolutionMsg, error) {
+
+	var msgs []*contractcourt.ResolutionMsg
+
+	err := kvdb.View(r.backend, func(tx kvdb.RTx) error {
+		resBucket := tx.ReadBucket(resBucketKey)
+		if resBucket == nil {
+			return nil
+		}
+
+		return resBucket.ForEach(func(k, v []byte) error {
+			kr := bytes.NewReader(k)
+			outKey := &CircuitKey{}
+			if err := outKey.Decode(kr); err != nil {
+				return err
+			}
+
+			vr := bytes.NewReader(v)
+			resMsg, err := deserializeResolutionMsg(vr)
+			if err != nil {
+				return err
+			}
+
+			// Set the CircuitKey values on the ResolutionMsg.
+			resMsg.SourceChan = outKey.ChanID
+			resMsg.HtlcIndex = outKey.HtlcID
+
+			msgs = append(msgs, resMsg)
+			return nil
+		})
+	}, func() {
+		msgs = nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return msgs, nil
+}
+
+// deleteResolutionMsg removes a ResolutionMsg with the passed-in CircuitKey.
+func (r *resolutionStore) deleteResolutionMsg(outKey *CircuitKey) error {
```

Comment:
> Do we also need a method to delete all resolution messages for a given channel?

I think so, unless we want to have a can at the _switch_level based on the circuit key, but that could very well be pushed into this layer, so it can be done in a _single_ database transaction.

Reply:
This is already done on switch startup